// @clickhouse/client/dist/connection/node_base_connection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeBaseConnection = void 0;
const client_common_1 = require("@clickhouse/client-common");
const crypto_1 = __importDefault(require("crypto"));
const utils_1 = require("../utils");
const stream_1 = require("./stream");
const socket_pool_1 = require("./socket_pool");
class NodeBaseConnection {
    params;
    agent;
    defaultAuthHeader;
    defaultHeaders;
    connectionId = crypto_1.default.randomUUID();
    socketPool;
    constructor(params, agent) {
        this.params = params;
        this.agent = agent;
        this.socketPool = new socket_pool_1.SocketPool(this.connectionId, this.params, this.createClientRequest.bind(this), this.agent);
        if (params.auth.type === 'Credentials') {
            this.defaultAuthHeader = `Basic ${Buffer.from(`${params.auth.username}:${params.auth.password}`).toString('base64')}`;
        }
        else if (params.auth.type === 'JWT') {
            this.defaultAuthHeader = `Bearer ${params.auth.access_token}`;
        }
        else {
            throw new Error(`Unknown auth type: ${params.auth.type}`);
        }
        this.defaultHeaders = {
            // Node.js HTTP agent, for some reason, does not set this on its own when KeepAlive is enabled
            Connection: this.params.keep_alive.enabled ? 'keep-alive' : 'close',
            'User-Agent': (0, utils_1.getUserAgent)(this.params.application_id),
        };
    }
    async ping(params) {
        const { log_writer, log_level } = this.params;
        const query_id = this.getQueryId(params.query_id);
        const { controller, controllerCleanup } = this.getAbortController(params);
        try {
            let result;
            if (params.select) {
                const searchParams = (0, client_common_1.toSearchParams)({
                    database: undefined,
                    query: PingQuery,
                    query_id,
                });
                result = await this.request({
                    query: PingQuery,
                    method: 'GET',
                    url: (0, client_common_1.transformUrl)({ url: this.params.url, searchParams }),
                    abort_signal: controller.signal,
                    headers: this.buildRequestHeaders(),
                    query_id,
                    log_writer,
                    log_level,
                }, 'Ping');
            }
            else {
                result = await this.request({
                    query: 'ping',
                    method: 'GET',
                    url: (0, client_common_1.transformUrl)({ url: this.params.url, pathname: '/ping' }),
                    abort_signal: controller.signal,
                    headers: this.buildRequestHeaders(),
                    query_id,
                    log_writer,
                    log_level,
                }, 'Ping');
            }
            await (0, stream_1.drainStreamInternal)({
                op: 'Ping',
                log_writer,
                query_id,
                log_level,
            }, result.stream);
            return { success: true };
        }
        catch (error) {
            // it is used to ensure that the outgoing request is terminated,
            // and we don't get unhandled error propagation later
            controller.abort('Ping failed');
            // not an error, as this might be semi-expected
            if (log_level <= client_common_1.ClickHouseLogLevel.WARN) {
                log_writer.warn({
                    message: this.httpRequestErrorMessage('Ping'),
                    err: error,
                    args: {
                        connection_id: this.connectionId,
                        query_id,
                    },
                });
            }
            return {
                success: false,
                error: error, // should NOT be propagated to the user
            };
        }
        finally {
            controllerCleanup();
        }
    }
    async query(params) {
        const { log_writer, log_level } = this.params;
        const query_id = this.getQueryId(params.query_id);
        const clickhouse_settings = (0, client_common_1.withHttpSettings)(params.clickhouse_settings, this.params.compression.decompress_response);
        const searchParams = (0, client_common_1.toSearchParams)({
            database: this.params.database,
            query_params: params.query_params,
            session_id: params.session_id,
            clickhouse_settings,
            query_id,
            role: params.role,
        });
        const { controller, controllerCleanup } = this.getAbortController(params);
        // allows enforcing the compression via the settings even if the client instance has it disabled
        const enableResponseCompression = clickhouse_settings.enable_http_compression === 1;
        try {
            const { response_headers, stream, http_status_code } = await this.request({
                method: 'POST',
                url: (0, client_common_1.transformUrl)({ url: this.params.url, searchParams }),
                body: params.query,
                abort_signal: controller.signal,
                enable_response_compression: enableResponseCompression,
                headers: this.buildRequestHeaders(params),
                query: params.query,
                query_id,
                log_writer,
                log_level,
            }, 'Query');
            return {
                stream,
                response_headers,
                query_id,
                http_status_code,
            };
        }
        catch (err) {
            controller.abort('Query HTTP request failed');
            this.logRequestError({
                op: 'Query',
                query_id: query_id,
                query_params: params,
                search_params: searchParams,
                err: err,
                extra_args: {
                    decompress_response: enableResponseCompression,
                    clickhouse_settings,
                },
            });
            throw err; // should be propagated to the user
        }
        finally {
            controllerCleanup();
        }
    }
    async insert(params) {
        const { log_writer, log_level } = this.params;
        const query_id = this.getQueryId(params.query_id);
        const searchParams = (0, client_common_1.toSearchParams)({
            database: this.params.database,
            clickhouse_settings: params.clickhouse_settings,
            query_params: params.query_params,
            query: params.query,
            session_id: params.session_id,
            role: params.role,
            query_id,
        });
        const { controller, controllerCleanup } = this.getAbortController(params);
        try {
            const { stream, summary, response_headers, http_status_code } = await this.request({
                method: 'POST',
                url: (0, client_common_1.transformUrl)({ url: this.params.url, searchParams }),
                body: params.values,
                abort_signal: controller.signal,
                enable_request_compression: this.params.compression.compress_request,
                parse_summary: true,
                headers: this.buildRequestHeaders(params),
                query: params.query,
                query_id,
                log_writer,
                log_level,
            }, 'Insert');
            await (0, stream_1.drainStreamInternal)({
                op: 'Insert',
                log_writer,
                query_id,
                log_level,
            }, stream);
            return { query_id, summary, response_headers, http_status_code };
        }
        catch (err) {
            controller.abort('Insert HTTP request failed');
            this.logRequestError({
                op: 'Insert',
                query_id: query_id,
                query_params: params,
                search_params: searchParams,
                err: err,
                extra_args: {
                    clickhouse_settings: params.clickhouse_settings ?? {},
                },
            });
            throw err; // should be propagated to the user
        }
        finally {
            controllerCleanup();
        }
    }
    async exec(params) {
        const query_id = this.getQueryId(params.query_id);
        return this.runExec({
            ...params,
            query_id,
            op: 'Exec',
        });
    }
    async command(params) {
        const { log_writer, log_level } = this.params;
        const query_id = this.getQueryId(params.query_id);
        const commandStartTime = Date.now();
        if (log_level <= client_common_1.ClickHouseLogLevel.TRACE) {
            log_writer.trace({
                message: 'Command: operation started',
                args: {
                    operation: 'Command',
                    connection_id: this.connectionId,
                    query_id,
                },
            });
        }
        const { stream, summary, response_headers } = await this.runExec({
            ...params,
            query_id,
            op: 'Command',
        });
        const runExecDuration = Date.now() - commandStartTime;
        if (log_level <= client_common_1.ClickHouseLogLevel.TRACE) {
            log_writer.trace({
                message: 'Command: runExec completed, starting stream drain',
                args: {
                    operation: 'Command',
                    connection_id: this.connectionId,
                    query_id,
                    runExec_duration_ms: runExecDuration,
                    stream_state: {
                        readable: stream.readable,
                        readableEnded: stream.readableEnded,
                        readableLength: stream.readableLength,
                    },
                },
            });
        }
        // ignore the response stream and release the socket immediately
        const drainStartTime = Date.now();
        await (0, stream_1.drainStreamInternal)({
            op: 'Command',
            log_writer,
            query_id,
            log_level,
        }, stream);
        if (log_level <= client_common_1.ClickHouseLogLevel.TRACE) {
            const drainDuration = Date.now() - drainStartTime;
            const totalDuration = Date.now() - commandStartTime;
            log_writer.trace({
                message: 'Command: operation completed',
                args: {
                    operation: 'Command',
                    connection_id: this.connectionId,
                    query_id,
                    drain_duration_ms: drainDuration,
                    total_duration_ms: totalDuration,
                },
            });
        }
        return { query_id, summary, response_headers };
    }
    async close() {
        if (this.agent !== undefined && this.agent.destroy !== undefined) {
            this.agent.destroy();
        }
    }
    defaultHeadersWithOverride(params) {
        return {
            // Custom HTTP headers from the client configuration
            ...(this.params.http_headers ?? {}),
            // Custom HTTP headers for this particular request; it will override the client configuration with the same keys
            ...(params?.http_headers ?? {}),
            // Includes the `Connection` + `User-Agent` headers which we do not allow to override
            // An appropriate `Authorization` header might be added later
            // It is not always required - see the TLS headers in `node_https_connection.ts`
            ...this.defaultHeaders,
        };
    }
    buildRequestHeaders(params) {
        const headers = this.defaultHeadersWithOverride(params);
        if ((0, client_common_1.isJWTAuth)(params?.auth)) {
            return {
                ...headers,
                Authorization: `Bearer ${params.auth.access_token}`,
            };
        }
        if (this.params.set_basic_auth_header) {
            if ((0, client_common_1.isCredentialsAuth)(params?.auth)) {
                return {
                    ...headers,
                    Authorization: `Basic ${Buffer.from(`${params.auth.username}:${params.auth.password}`).toString('base64')}`,
                };
            }
            else {
                return {
                    ...headers,
                    Authorization: this.defaultAuthHeader,
                };
            }
        }
        return {
            ...headers,
        };
    }
    getQueryId(query_id) {
        return query_id || crypto_1.default.randomUUID();
    }
    // a wrapper over the user's Signal to terminate the failed requests
    getAbortController(params) {
        const controller = new AbortController();
        function onAbort() {
            controller.abort();
        }
        params.abort_signal?.addEventListener('abort', onAbort);
        return {
            controller,
            controllerCleanup: () => {
                params.abort_signal?.removeEventListener('abort', onAbort);
            },
        };
    }
    logRequestError({ op, err, query_id, query_params, extra_args, }) {
        if (this.params.log_level <= client_common_1.ClickHouseLogLevel.ERROR) {
            this.params.log_writer.error({
                message: this.httpRequestErrorMessage(op),
                err: err,
                args: {
                    operation: op,
                    connection_id: this.connectionId,
                    query_id,
                    with_abort_signal: query_params.abort_signal !== undefined,
                    session_id: query_params.session_id,
                    ...extra_args,
                },
            });
        }
    }
    httpRequestErrorMessage(op) {
        return `${op}: HTTP request error.`;
    }
    async runExec(params) {
        const { log_writer, log_level } = this.params;
        const query_id = params.query_id;
        const sendQueryInParams = params.values !== undefined;
        const clickhouse_settings = (0, client_common_1.withHttpSettings)(params.clickhouse_settings, this.params.compression.decompress_response);
        const toSearchParamsOptions = {
            query: sendQueryInParams ? params.query : undefined,
            database: this.params.database,
            query_params: params.query_params,
            session_id: params.session_id,
            role: params.role,
            clickhouse_settings,
            query_id,
        };
        const searchParams = (0, client_common_1.toSearchParams)(toSearchParamsOptions);
        const { controller, controllerCleanup } = this.getAbortController(params);
        const tryDecompressResponseStream = params.op === 'Exec'
            ? // allows disabling stream decompression for the `Exec` operation only
                (params.decompress_response_stream ??
                    this.params.compression.decompress_response)
            : // there is nothing useful in the response stream for the `Command` operation,
                // and it is immediately destroyed; never decompress it
                false;
        const ignoreErrorResponse = params.ignore_error_response ?? false;
        try {
            const { stream, summary, response_headers, http_status_code } = await this.request({
                method: 'POST',
                url: (0, client_common_1.transformUrl)({ url: this.params.url, searchParams }),
                body: sendQueryInParams ? params.values : params.query,
                abort_signal: controller.signal,
                parse_summary: true,
                enable_request_compression: this.params.compression.compress_request,
                enable_response_compression: this.params.compression.decompress_response,
                try_decompress_response_stream: tryDecompressResponseStream,
                ignore_error_response: ignoreErrorResponse,
                headers: this.buildRequestHeaders(params),
                query: params.query,
                query_id,
                log_writer,
                log_level,
            }, params.op);
            return {
                stream,
                query_id,
                summary,
                response_headers,
                http_status_code,
            };
        }
        catch (err) {
            controller.abort(`${params.op} HTTP request failed`);
            this.logRequestError({
                op: params.op,
                query_id: query_id,
                query_params: params,
                search_params: searchParams,
                err: err,
                extra_args: {
                    clickhouse_settings: params.clickhouse_settings ?? {},
                },
            });
            throw err; // should be propagated to the user
        }
        finally {
            controllerCleanup();
        }
    }
    async request(params, op) {
        return this.socketPool.request(params, op);
    }
}
exports.NodeBaseConnection = NodeBaseConnection;
const PingQuery = `SELECT 'ping'`;
